Refactor statistics to use logging table with pagination and incremental refresh

System625 opened the pull request:
## Summary
Refactored review statistics backend to use the MediaWiki `logging` table instead of `flaggedrevs` table, enabling pagination support for larger datasets (30-90 days) and incremental refresh functionality. Closes issue #81 

## Changes Made

### Backend Refactoring
- **Moved statistics logic to separate file** (`app/reviews/services/statistics.py`) to prevent merge conflicts
- **Switched from flaggedrevs to logging table** query for full historical data including deleted pages
- **Added pagination support** to handle Superset's 10k row limit (fetches multiple batches automatically)
- **Implemented incremental refresh** using `max_log_id` tracking to fetch only new data
- **Added `max_log_id` field** to `ReviewStatisticsMetadata` model for continuation support

### API Changes
- **Updated `/api/wikis/<pk>/statistics/refresh/`** - Now performs incremental refresh (fast)
- **Added `/api/wikis/<pk>/statistics/clear/`** - Full reload with configurable days (slow but complete)
- Changed `fetch_review_statistics()` parameter from `limit` to `days`

### UI Improvements
- **"Refresh (incremental)"** button - Fetches only new reviews since last refresh (~10-20 seconds)
- **"Clear & Reload"** button - Clears cache and loads 30 days of fresh data (~1-2 minutes)
- **Independent loading states** - Each button shows its own loading spinner
- Added helpful tooltips and Font Awesome icons for better UX

### Management Commands
- **`compare_statistics_queries`** - Validates that logging table query produces similar results to flaggedrevs query
- **`refresh_statistics`** - Incrementally refresh statistics for one or all wikis
- **`load_statistics`** - Load full statistics with custom time period (e.g., 90 days)

## Validation
✅ Comparison command confirms logging table query is safe to use (67.9% revision overlap, identical median delay)  
✅ All 173 tests passing  
✅ UI tested - separate loading states work correctly  
✅ Incremental refresh successfully fetches only new data  

## Benefits
- **Supports larger datasets**: Can now load 30-90 days of data via pagination (previously limited to 10k rows)
- **Faster updates**: Incremental refresh fetches only new data instead of reloading everything
- **Better UX**: Clear distinction between "refresh" (incremental) and "clear & reload" (full)
- **More complete data**: Logging table includes deleted pages and historical review actions

## Screenshots
<img width="1315" height="338" alt="image" src="https://github.com/user-attachments/assets/7da9e0bd-69e5-4426-b1c5-6fac65513672" />

## Testing
```bash
# Run tests
python manage.py test

# Compare queries to validate
python manage.py compare_statistics_queries --wiki fi --limit 1000

# Test incremental refresh
python manage.py refresh_statistics --wiki fi

# Test full reload
python manage.py load_statistics --wiki fi --days 30

zache-fi commented on app/reviews/services/statistics.py:
```diff
@@ -0,0 +1,654 @@
+            logger.info("Batch %d: fetched %d records", batches_fetched, batch_count)
+
+            # If we got less than 10k records, we've reached the end
+            if batch_count < 10000:
```
because LIMIT 10000 is in inner SQL query which is joined to other tables there is possibility that later parts of the query reduces the number of rows to be lower than 10000 and it will newer have 10000 rows even if there is more data available.

You can test it by trying load data for 30 days from dewiki for example

System625 commented:
Hi @zache-fi , apologies for overlooking big bugs like that
I have fixed all issues from review:

1.  Changed ORDER BY from DESC to ASC (both inner and outer queries)
2.  Fixed pagination logic - checks if max_log_id advances instead of row count
3.  Updated comparison command to filter both queries to same time period

Test results:
- Comparison: 99.5% overlap, 0.0% difference in counts/delays
- Finnish Wikipedia: 5,163 records in 2 batches
- German Wikipedia: 34,508 records in 6 batches
- Pagination correctly stops when max_log_id doesn't advance

Ready for review!

zache-fi commented on app/reviews/services/statistics.py:
```diff
@@ -0,0 +1,664 @@
+            previous_max_log_id = current_max_log_id
+
+            # Safety limit: don't fetch more than 10 batches (100k records)
+            if batches_fetched >= 10:
```
this limit comes too fast.

System625 commented:
Hello @zache-fi I have fixed the comparison command as requested:
- Old query now uses same timestamp range as new query (gets min/max from new query results first)
- Changed ORDER BY from fr_rev_id DESC to fr_timestamp ASC
- Added mandatory fr_timestamp filter clauses to prevent timeout

Tested with Finnish Wikipedia (7 days, 5000 limit):
- 884 records from both queries
- 99.5% overlap
- 0% count difference
- Identical delay statistics

For the enhancements and refactor, I think it's best I do that in a new PR to not complicate this so you can go ahead and merge this. Once merged, I will open a new PR to address the enhancement

zache-fi commented:
I did the merge now. I was just checking if there is something else which should be merged first as merging tends to generate merge conflicts. Feel free to write followup github issue  tickets for next tasks.
